### DISCUSSION TITLE
How to output custom json log message

### DISCUSSION THREAD

**Discussion by missedone** [keycloak/keycloak#29091] [Q&A] [answered]
hi folks
i'd like to output the keycloak logs in the custom json format, i know keycloak uses quarkus which use jboss-logging, but the problem with jboss-logging is it doesn't support escape quotes or encode multiline string in the json message https://www.keycloak.org/server/logging#_configuring_the_console_log_format

how can i make keycloak use log4j2? or do you see any other option?

thanks

**Comment by codespearhead** [comment, marked as answer]
Are you saying that `KC_LOG_CONSOLE_OUTPUT=json` does not work?

**Comment by missedone** [comment]
hi, `KC_LOG_CONSOLE_OUTPUT=json` gives the default json format while i want to custom the json message format

**Comment by missedone** [comment]
in my case, i'd like to get the log output in format of `{"applog":{"EventTime":"%d{yyyy-MM-dd HH:mm:ss,SSS}","meta":{"level":"%-5p","logger":"%c","thread":"%t"},"event":"%s"}}`

**Reply by codespearhead** [reply]
I see. If a [custom format](https://www.keycloak.org/server/logging#:~:text=Configuring%20the%20console) via [KC_LOG_CONSOLE_FORMAT](https://www.keycloak.org/server/all-config?q=log-console-format) as well as using [Quarkus directly](https://github.com/keycloak/keycloak/discussions/10201#discussioncomment-2178679) does not work, the the only way out I can think of is via [file log parsing](https://github.com/keycloak/keycloak/blob/a6e2ab5523f3b9fc0a4283b9a6d4f48b3d97b2a6/docs/documentation/release_notes/topics/24_0_0.adoc?plain=1#L443).

**Reply by missedone** [reply]
I tried KC_LOG_CONSOLE_FORMAT with the above json format, but the log didn’t show up at all on console, I guess it could be because of the braces,  the simple plain format without braces worked

**Reply by codespearhead** [reply]
Try it like this:
 
```bash
docker run -p 8080:8080 \
-e KEYCLOAK_ADMIN=admin \
-e KC_LOG_CONSOLE_OUTPUT=json \
-e KEYCLOAK_ADMIN_PASSWORD=admin \
-e KC_LOG_CONSOLE_FORMAT='{"applog":{"EventTime":"%d{yyyy-MM-dd HH:mm:ss,SSS}","meta":{"level":"%-5p","logger":"%c","thread":"%t"},"event":"%s"}}' \
quay.io/keycloak/keycloak:24.0.3 start-dev
```

It seems to be working on my end.

**Comment by missedone** [comment]
@codespearhead , interesting, it worked now with `KC_LOG_CONSOLE_FORMAT`, not sure what wrong before.

thanks

**Comment by missedone** [comment]
hmm, with more test, the solution is not perfect because it can't handle/escape the double quotes, new lines and other special characters in the log message.
Ex. below is not a valid json: 
```
{"applog":{"EventTime":"2024-04-29 21:06:25,525","meta":{"level":"WARN ","logger":"org.keycloak.events","thread":"executor-thread-94"},"event":"type="INTROSPECT_TOKEN_ERROR", realmId="5384fc72-1822-438d-96ee-a851b3f93e5f", clientId="example-svc", userId="null", ipAddress="100.126.62.50", error="token_introspection_failed", reason="Could not find valid user session for session_state = 4b57f130-529a-4b2d-bde6-c09fa7927b47", client_auth_method="client-secret""}}
```

**Reply by codespearhead** [reply]
What action did you perform on the server to get the above output?

Either way, I [updated my previous answer](https://github.com/keycloak/keycloak/discussions/29091#discussioncomment-9248185) to add the missing `KC_LOG_CONSOLE_OUTPUT=json`. I suggest trying it again.

**Reply by missedone** [reply]
@codespearhead thanks for the update, if `KC_LOG_CONSOLE_OUTPUT=json` is set, it use the default json format rather than my custom json format defined in `KC_LOG_CONSOLE_FORMAT`

**Reply by missedone** [reply]
anyway, i will take your suggestion as it's the most viable option.

at the server side, we have filebeat collect the logs, so i will have dedicate log parse pipeline in filebeat/ES for keycloak.
